ir/state.cpp

```diff
@@ -143,10 +238,24 @@ static expr not_poison_except_padding(const Type &ty, const expr &np) {
   return result;
 }
 
-const StateValue& State::getAndAddPoisonUB(const Value &val) {
-  auto &v = (*this)[val];
-  if (!analysis.non_poison_vals.insert(&val).second)
-    return v;
+const StateValue&
+State::getAndAddPoisonUB(const Value &val, bool undef_ub_too) {
+  auto &sv = (*this)[val];
+
+  if (!analysis.non_poison_vals.insert(&val).second && !undef_ub_too)
+    return sv;
+
+  expr v = sv.value;
+
+  if (undef_ub_too) {
+    auto inserted = analysis.non_undef_vals.find(&val);
+    if (inserted != analysis.non_undef_vals.end()) {
+      v = inserted->second;
+    } else {
+      v = strip_undef_and_add_ub(*this, val, v);
+      analysis.non_undef_vals[&val] = v;
+    }
+  }
 
   // mark all operands of val as non-poison if they propagate poison
```

Comment:
we should skip this bit (up to addUb below) if we had already marked the value as non-poison before.

Reply:
Added a guard